### PR TITLE
Make `qml.QutritBasisStatePreparation` JIT compatible

### PR DESCRIPTION
**Pull request by willjmax**
This PR makes `qml.QutritBasisStatePreparation` JIT compatible. Previously, the template used the `qml.TShift` operator in the decomposition. However, this approach requires non-jittable control flow on the input. To make the template jittable the decomposition is changed to return a `qml.ops.QutritUnitary` representing a `TShift` operator raised to some power. This change requires adding `matrix_power` and `eigh` to the multi dispatch. This PR addresses sc-70863.

### CODE REVIEW

File: tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py [hunk 1]
```diff
@@ -128,6 +131,7 @@ def circuit(state, obs):
         obs = np.array([[1, 0, 0], [0, 2, 0], [0, 0, 3]])
         output_state = [x - 1 for x in circuit(basis_state, obs)]
 
+        print(output_state, target_state)
```
Comment (KetpuntoG) [inline tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py:134]:
```suggestion
```

File: pennylane/math/multi_dispatch.py [hunk 2]
```diff
@@ -315,6 +315,44 @@ def matmul(tensor1, tensor2, like=None):
     return ar.numpy.matmul(tensor1, tensor2, like=like)
 
 
+@multi_dispatch(argnum=[0, 1])
```
Comment (KetpuntoG) [inline pennylane/math/multi_dispatch.py:318]: With each new function added, we should include a test in `test/math/test_multi_dispactch`

File: pennylane/templates/state_preparations/basis_qutrit.py [hunk 3]
```diff
@@ -112,7 +113,10 @@ def compute_decomposition(basis_state, wires):  # pylint: disable=arguments-diff
...
 
         op_list = []
+        tshift = qml.math.array([[0, 0, 1], [1, 0, 0], [0, 1, 0]])
         for wire, state in zip(wires, basis_state):
-            for _ in range(0, state):
-                op_list.append(qml.TShift(wire))
+            mat = qml.math.matrix_power(tshift, state)
+            op = qml.ops.QutritUnitary(mat, wires=wire)
+            op_list.append(op)
+
```
Comment (KetpuntoG) [inline pennylane/templates/state_preparations/basis_qutrit.py:121]: to avoid QubitUnitary (since it is not hardware compatible), you could use this:

`qml.pow(qml.TShift(wires = wire),state)`

qml.pow returns an operator, so the idea is the same without access the matrix. 

Note that if this works, you may not need to add matrix power in the `dispach.py` file

Comment (willjmax) [reply pennylane/templates/state_preparations/basis_qutrit.py:121]: The issue is that the `pow` method for `qml.TShift` is not jittable. It uses the `pow` method from `Operator` which has control flow to check if the exponent is `0` or `1`.

File: pennylane/ops/qutrit/observables.py [hunk 4]
```diff
@@ -110,13 +110,18 @@ def eigendecomposition(self):
...
-        if Hkey not in THermitian._eigs:
-            w, U = np.linalg.eigh(Hmat)
-            THermitian._eigs[Hkey] = {"eigvec": U, "eigval": w}
 
-        return THermitian._eigs[Hkey]
+        if not qml.math.is_abstract(Hmat):
+            Hmat = qml.math.to_numpy(Hmat)
+            Hkey = tuple(Hmat.flatten().tolist())
+            if Hkey not in THermitian._eigs:
+                w, U = qml.math.eigh(Hmat)
```
Comment (obliviateandsurrender) [inline pennylane/ops/qutrit/observables.py:118]: How come `qml.math.linalg.eigh` doesn't work? 🤔

File: pennylane/templates/state_preparations/basis_qutrit.py [hunk 5]
```diff
@@ -112,7 +113,10 @@ def compute_decomposition(basis_state, wires):  # pylint: disable=arguments-diff
         """
 
         op_list = []
+        tshift = qml.math.array([[0, 0, 1], [1, 0, 0], [0, 1, 0]])
         for wire, state in zip(wires, basis_state):
-            for _ in range(0, state):
-                op_list.append(qml.TShift(wire))
+            mat = qml.math.matrix_power(tshift, state)
```
Comment (obliviateandsurrender) [inline pennylane/templates/state_preparations/basis_qutrit.py:118]: We cannot use `qml.math.linalg.matrix_power`? 🤔

File: pennylane/templates/state_preparations/basis_qutrit.py [hunk 6]
```diff
@@ -112,7 +113,10 @@ def compute_decomposition(basis_state, wires):  # pylint: disable=arguments-diff
         """
 
         op_list = []
+        tshift = qml.math.array([[0, 0, 1], [1, 0, 0], [0, 1, 0]])
```
Comment (obliviateandsurrender) [inline pennylane/templates/state_preparations/basis_qutrit.py:116]: Does this scale worse or better than the default case? If worse, then should we limit this only to `jit` case (and use the default behaviour otherwise)?

**Review by obliviateandsurrender — COMMENTED** [review]
Thanks @willjmax! Some minor queries, before I dig in deeper.

File: pennylane/ops/qutrit/observables.py (hunk 4 above)
Comment (willjmax) [reply pennylane/ops/qutrit/observables.py:118]: Apparently it does, so I'm removing it from multi dispatch.

File: pennylane/templates/state_preparations/basis_qutrit.py (hunk 6 above)
Comment (willjmax) [reply pennylane/templates/state_preparations/basis_qutrit.py:116]: Good suggestion. I'm changing it to default to the old decomposition, and only decompose to `qml.QutritUnitary` when jit is enabled.

File: pennylane/ops/qubit/observables.py [hunk 7]
```diff
@@ -91,6 +91,9 @@ def __init__(self, A: TensorLike, wires: WiresLike, id: Optional[str] = None):
     @staticmethod
     def _validate_input(A: TensorLike, expected_mx_shape: Optional[int] = None):
         """Validate the input matrix."""
+        if qml.math.is_abstract(A):
+            return
```
Comment (soranjh) [inline pennylane/ops/qubit/observables.py:95]: Why this is needed now?

File: pennylane/templates/state_preparations/basis_qutrit.py [hunk 8]
```diff
@@ -15,6 +15,7 @@
 Contains the BasisStatePreparation template.
 """
 
+from itertools import product
```
Comment (soranjh) [inline pennylane/templates/state_preparations/basis_qutrit.py:18]: Please remove the import if not used.

File: pennylane/templates/state_preparations/basis_qutrit.py [hunk 9]
```diff
@@ -112,7 +114,18 @@ def compute_decomposition(basis_state, wires):  # pylint: disable=arguments-diff
         """
 
         op_list = []
+
+        if qml.math.is_abstract(basis_state):
+            tshift = qml.math.array([[0, 0, 1], [1, 0, 0], [0, 1, 0]])
+            for wire, state in zip(wires, basis_state):
+                mat = qml.math.linalg.matrix_power(tshift, state)
+                op = qml.ops.QutritUnitary(mat, wires=wire)
```
Comment (soranjh) [inline pennylane/templates/state_preparations/basis_qutrit.py:122]: So the workflow for computing gradients has a different decomposition which is not hardware-friendly?

File: pennylane/ops/qubit/observables.py (hunk 7 above)
Comment (willjmax) [reply pennylane/ops/qubit/observables.py:95]: At some point when the qnode is executed `Hermitian.compute_matrix` is called. The `_validate_input` function to determine if the input is Hermitian is not jittable, so we check if the input is abstract and skip the validation if it is.

**Review by soranjh — APPROVED** [review]

File: tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py [hunk 10]
```diff
@@ -102,33 +105,23 @@ def circuit(obs):
...
 
         circuit = jax.jit(circuit)
 
-        obs = np.array([[1, 0, 0], [0, 2, 0], [0, 0, 3]])
-        output_state = [x - 1 for x in circuit(basis_state, obs)]
+        basis_state = qml.math.array([2], like="jax")
+        output_state = circuit(basis_state)
 
-        assert np.allclose(output_state, target_state, atol=tol, rtol=0)
+        assert qml.math.allclose(output_state, [0, 0, 1])
```
Comment (soranjh) [inline tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py:124]: Just double checking, is this state obtained with the new decomposition using TRY gates instead of TShift gates? Or a matrix is used?

**Review by KetpuntoG — COMMENTED** [review]
Nice! 😄

Just one comment. The following example does not work with `jax.jit`. For some reason `qml.sample` fails. I have tested `default.qubit` with `BasisState` to see if it should fail but it works correctly there.

```python
n = 2

@jax.jit
@qml.qnode(qml.device("default.qutrit", wires = n, shots = 1))
def circuit(state):

    qml.QutritBasisStatePreparation(state, wires = range(n))
    
    return qml.sample(wires = range(n))

state = jax.numpy.array([1,1])
circuit(state)
```

Otherwise, the PR looks good to me :)

File: pennylane/templates/state_preparations/basis_qutrit.py [hunk 11]
```diff
@@ -112,7 +115,23 @@ def compute_decomposition(basis_state, wires):  # pylint: disable=arguments-diff
...
+        if qml.math.is_abstract(basis_state):
+            for wire, state in zip(wires, basis_state):
+                op_list.extend(
+                    [
+                        qml.TRY(state * (2 - state) * np.pi, wires=wire, subspace=(0, 1)),
+                        qml.TRY(state * (1 - state) * np.pi / 2, wires=wire, subspace=(0, 2)),
+                        qml.TRZ((-2 * state + 3) * state * np.pi, wires=wire, subspace=(0, 2)),
+                        qml.TRY(state * (2 - state) * np.pi, wires=wire, subspace=(0, 2)),
+                        qml.TRY(state * (1 - state) * np.pi / 2, wires=wire, subspace=(0, 1)),
+                        qml.TRZ(-(7 * state - 10) * state * np.pi, wires=wire, subspace=(0, 2)),
```
Comment (obliviateandsurrender) [inline pennylane/templates/state_preparations/basis_qutrit.py:128]: Is there a resource for this decomposition?

Comment (willjmax) [reply pennylane/templates/state_preparations/basis_qutrit.py:128]: Nope, it came straight from @KetpuntoG's :brain:. We should add extra tests to ensure the JIT decomposition is correct.

File: tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py [hunk 12]
```diff
@@ -102,33 +105,57 @@ def circuit(obs):
...
+        import jax.numpy as jnp
+
+        tshift = jnp.array([[0, 0, 1], [1, 0, 0], [0, 1, 0]])
+
+        jit_decomp = jax.jit(qml.QutritBasisStatePreparation.compute_decomposition)
+        jit_decomp(jnp.array([1]), wires=[0])
+
+        for i in range(3):
+            decomp = jit_decomp(jnp.array([i]), wires=[0])
+            matrix = qml.matrix(qml.prod(*decomp[::-1]))
```
Comment (obliviateandsurrender) [inline tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py:157]: Could we also test the decompositions in terms of `operators`?

Comment (soranjh) [reply tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py:157]: Yes, we need to test two things:
1. The jit decomposition is equivalent to the list of TR gates as expected. You can manually list the TR gates and compare.
2. The new decomposition in terms of TR gates is correct. For that we can compute the matrix representation of the decomposition and compare it with a reference matrix.

Comment (willjmax) [reply tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py:152]: I added a test for the first thing, the second is already handled by my previous test.

File: tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py [hunk 13]
```diff
@@ -102,33 +102,79 @@ def circuit(obs):
...
+    def test_decomposition_pl_gates_jax_jit(self):
+        """Tests that the decomposition gates are correct when JIT compiled."""
+        import jax
+        import jax.numpy as jnp
+
+        jit_decomp = jax.jit(qml.QutritBasisStatePreparation.compute_decomposition)
+        jit_decomp(jnp.array([1]), wires=jnp.array([0]))
+
+        def op_list(state, wire):
+            return [
```
Comment (soranjh) [inline tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py:165]: It better to be a list: `op_list = [...]`

**Review by KetpuntoG — APPROVED** [review]
Looks good! 👌

File: tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py [hunk 14]
```diff
@@ -102,33 +102,79 @@ def circuit(obs):
...
+                qml.TRZ((-2 * state + 3) * state * np.pi, wires=wire, subspace=(0, 2)),
+                qml.TRY(state * (2 - state) * np.pi, wires=wire, subspace=(0, 2)),
+                qml.TRY(state * (1 - state) * np.pi / 2, wires=wire, subspace=(0, 1)),
+                qml.TRZ(-(7 * state - 10) * state * np.pi, wires=wire, subspace=(0, 2)),
+            ]
+
+        for i in range(3):
+            decomp = jit_decomp(jnp.array([i]), wires=jnp.array([0]))
+            for op1, op2 in zip(decomp, op_list(i, [0])):
+                assert op1 == op2
```
Comment (obliviateandsurrender) [inline tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py:177]:
```suggestion
                qml.assert_equal(op1, op2)
```

**Review by obliviateandsurrender — APPROVED** [review]
[Non blocker] could combine `matrix` and `pl_gates` tests into one `decomposition` test, otherwise looks good with just one minor thing.

**Comment by willjmax** [comment]
> I think qutrit basis state prep test in `templates/state_preparations/basis_qutrit.py` should be made to `xfail` for the CI to pass.

No, this was due to a bug that I caused. I pushed a fix.
